### PR TITLE
[thci] enhance THCI for 1.2 Certification

### PR DESCRIPTION
**Pull request by simonlingoogle**
This commit enhances THCI for 1.2 Certification:
- OpenThread.py: 1.2 non-BR
- OpenThread_BR.py: 1.2 BR and Host (`otbr-agent` solution)

This commit should also makes the THCI works for both TH1.1 and TH1.2. 

TODO:
- [x] Basic TestHarness 1.2 regression
  - [x] TC involving 1.1 device
  - ~~TC involving external Commissioners (Not supported)~~

### CODE REVIEW

**Comment by romacdon** [comment]
Hi @simonlingoogle - I'm curious about the status of this PR.  Any thoughts on when this will be merged.  Also, do you anticipate any other changes to things like log messages or CLI for integration with the TH?

**Comment by simonlingoogle** [comment]
> Hi @simonlingoogle - I'm curious about the status of this PR. Any thoughts on when this will be merged. Also, do you anticipate any other changes to things like log messages or CLI for integration with the TH?

Hi @romacdon , thanks for asking. 

This PR will be ready for merging once the LGTM alerts are resolved. 
The THCI is actually tested on TH, and works for TH1.2, even though many TCs are still failing (mostly fail not because of THCI). 
I have fixed most THCI issues I found when running TH, which makes this PR a good step towards a full workable THCI for TH1.2.

File: tools/harness-thci/OpenThread.py [hunk 1]
```diff
@@ -3003,6 +3220,214 @@ def setCSLsuspension(self, suspend):
...
+                else:
+                    raise
+
+        return True
+
+    @API
+    def registerMulticast(self, sAddr='ff04::1234:777a:1'):
+        """subscribe to the given ipv6 address (sAddr) in interface and send MLR.req OTA
+
+        note: workaround agreed before finial decision discussed in the DEV-1819.
```
Comment (librasungirl) [inline tools/harness-thci/OpenThread.py:3341]: not sure is it still required given fake option via `send_udp()`?

File: tools/harness-thci/OpenThread.py [hunk 2]
```diff
@@ -3003,6 +3220,214 @@ def setCSLsuspension(self, suspend):
...
+    @API
+    def setLeaderWeight(self, iWeight=72):
+        self.__executeCommand('leaderweight %d' % iWeight)
+
+    def __discoverDeviceCapability(self):
+        """Discover device capability according to version"""
+        self.DeviceCapability = DevCapb.NotSpecified
+
+        if self.IsBorderRouter:
+            self.DeviceCapability = DevCapb.C_BBR | DevCapb.C_Host | DevCapb.C_Comm
```
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply tools/harness-thci/OpenThread.py:3419]: For Host, we are missing `mldv2_query` and `mdns_query` APIs. Should be fine to remove Host.

File: tools/harness-thci/OpenThread.py [hunk 3]
```diff
@@ -3003,6 +3220,214 @@ def setCSLsuspension(self, suspend):
...
+
+        if mliid is not None:
+            mliid = mliid.replace(':', '')
+            cmd += ' %s' % mliid
+
+        self.__executeCommand(cmd)
+
+    @API
+    def getDUA(self):
+        dua = self.getGUA('fd00:7d03')
```
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply tools/harness-thci/OpenThread.py:3246]: The domain prefix is always `fd00:7d03:7d03:7d03::/64`. In the TCs that do not use this domain prefix, there is no domain prefix at all. So, a DUA always startswith `fd00:7d03:7d03:7d03`.

File: tools/harness-thci/OpenThread.py [hunk 4]
```diff
@@ -3003,6 +3220,214 @@ def setCSLsuspension(self, suspend):
...
+        return mliid
+
+    def __setMlIid(self, sMlIid):
+        """Set the Mesh Local IID before Thread Starts."""
+        assert ':' not in sMlIid
+        cmd = 'mliid %s' % sMlIid
+        self.__executeCommand(cmd)
+
+    @API
+    def registerDUA(self, destAddr='', sMleId='', sAddr=''):
```
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply tools/harness-thci/OpenThread.py:3276]: I have checked. `registerDUA` is still used by multiple TCs.

File: tools/harness-thci/OpenThread.py [hunk 5]
```diff
@@ -3003,6 +3220,214 @@ def setCSLsuspension(self, suspend):
...
+        print(cmd)
+        return self.__executeCommand(cmd)[-1] == 'Done'
+
+    @API
+    def config_next_mlr_status_rsp(self, status_code):
+        cmd = 'bbr mgmt mlr response %d' % status_code
+        return self.__executeCommand(cmd)[-1] == 'Done'
+
+    @API
+    def setMLRtimeout(self, iMsecs):
```
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply tools/harness-thci/OpenThread.py:3312]: Good question. The unit is actually `seconds` despite the deceptive argument name. 
I leave the argument name unchanged to make it same as the TestHarness THCI abstract class.

File: tools/harness-thci/OpenThread.py (hunk 1 above)
Comment (simonlingoogle) [reply tools/harness-thci/OpenThread.py:3341]: `registerMulticast` actually subscribes to the MA, not just send fake MLR.req.

File: tools/harness-thci/OpenThread.py [hunk 6]
```diff
@@ -2963,23 +3176,27 @@ def setTxPower(self, level):
     def sendUdp(self, destination, port, payload='hello'):
         self.log('call sendUdp')
         assert payload is not None, 'payload should not be none'
-        cmd1 = 'udp open'
-        print(cmd1)
-        cmd2 = 'udp send %s %d %s' % (destination, port, payload)
-        print(cmd2)
-        return self.__executeCommand(cmd1)[-1] == 'Done' and self.__executeCommand(cmd2)[-1] == 'Done'
+        cmd = 'udp send %s %d %s' % (destination, port, payload)
```
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply tools/harness-thci/OpenThread.py:3179]: fixed

File: tools/harness-thci/OpenThread.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (librasungirl) [reply tools/harness-thci/OpenThread.py:3276]: What is `registerDUA` used for in these TCs? 
Note that there is no `dua reg`  CLI command in public repo, so this API could not serve to send DUA.req.

[the remark this answers is not in the document]
Comment (simonlingoogle) [reply tools/harness-thci/OpenThread.py:3276]: Good point!

I think only these code are actually useful in this method:
```
        if sAddr != '' and destAddr == '' and sMleId == '':
            self.__setDUA(sAddr)
            return
```

We can remove arguments `destAddr='', sMleId=''` and leaving only `sAddr`

[the remark this answers is not in the document]
Comment (simonlingoogle) [reply tools/harness-thci/OpenThread.py]: Fixed

File: tools/harness-thci/OpenThread.py (hunk 1 above)
Comment (simonlingoogle) [reply tools/harness-thci/OpenThread.py]: Yes, we can remove the note.

File: tools/harness-thci/OpenThread.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply tools/harness-thci/OpenThread.py:3292]: There is no use case to use another domain. Prefer to use less coding if possible.

File: tools/harness-thci/OpenThread.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (simonlingoogle) [reply tools/harness-thci/OpenThread.py]: Removed Host cap

**Review by wgtdkp — APPROVED** [review]
Thanks for adding the extrnal commissioner APIs :+1: !

**Review by jwhui — APPROVED** [review]

**Comment by romacdon** [comment]
Thanks @simonlingoogle for all the work on these changes!

**Pull request merged**